qa/acceptance-tests/src/test/java/io/camunda/it/client/UsersByRoleIntegrationTest.java

```diff
@@ -0,0 +1,430 @@
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("username must not be null");
+  }
+
+  @Test
+  void shouldRejectAssignIfUsernameIsEmpty() {
+    assertThatThrownBy(
+            () -> camundaClient.newAssignUserToRoleCommand("roleId").username("").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("username must not be empty");
+  }
+
+  @Test
+  void shouldRejectAssignIfRoleIdIsNull() {
+    assertThatThrownBy(
+            () -> camundaClient.newAssignUserToRoleCommand(null).username("username").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("roleId must not be null");
+  }
+
+  @Test
+  void shouldRejectAssignIfRoleIdIsEmpty() {
+    assertThatThrownBy(
+            () -> camundaClient.newAssignUserToRoleCommand("").username("username").send().join())
+        .isInstanceOf(IllegalArgumentException.class)
+        .hasMessageContaining("roleId must not be empty");
+  }
+
+  @Test
+  void shouldUnassignUsersFromRoleOnRoleDeletion() {
+    // given
+    final var roleId = Strings.newRandomValidIdentityId();
+    final var username1 = Strings.newRandomValidUsername();
+    final var username2 = Strings.newRandomValidUsername();
+    createUser(username1);
+    createUser(username2);
+    createRole(roleId, "roleName", "roleDesc");
+
+    camundaClient.newAssignUserToRoleCommand(roleId).username(username1).send().join();
+    camundaClient.newAssignUserToRoleCommand(roleId).username(username2).send().join();
+
+    Awaitility.await("Users should appear in role user search")
+        .atMost(TIMEOUT_DATA_AVAILABILITY)
+        .untilAsserted(
+            () -> {
+              final var users =
+                  camundaClient.newUsersByRoleSearchRequest(roleId).send().join().items();
+              assertThat(users)
+                  .extracting(User::getUsername)
+                  .containsExactlyInAnyOrder(username1, username2);
+            });
+
+    // when
+    camundaClient.newDeleteRoleCommand(roleId).send().join();
+
+    // then
+    Awaitility.await("Users should be unassigned after role deletion")
+        .atMost(TIMEOUT_DATA_AVAILABILITY)
+        .untilAsserted(
+            () -> {
+              final var users =
+                  camundaClient.newUsersByRoleSearchRequest(roleId).send().join().items();
+              assertThat(users).isEmpty();
+            });
+  }
+
+  // unassign tests
+
+  @Test
+  void shouldUnassignUserFromRole() {
+    // given
+    final var username = Strings.newRandomValidUsername();
+    final var roleId = Strings.newRandomValidIdentityId();
+    createUser(username);
+    createRole(roleId, "roleName", "roleDesc");
+    camundaClient.newAssignUserToRoleCommand(roleId).username(username).send().join();
+
+    // when
+    camundaClient.newUnassignUserFromRoleCommand(roleId).username(username).send().join();
```

Comment:
🔧 To ensure the test functions as it should be we need an await we assign the role to the user. Since secondary storage has a delay it could now happen that we assign the role to the user, don't actually unassign it and that the assertion still passes as secondary storage hasn't updated yet.